### ISSUE TITLE
implement a PEP 528 windows console (io._WindowsConsoleIO)

### ISSUE THREAD

**Issue by gitlab-importer** [pypy/pypy#2999]
In Heptapod by bitbucket_importer on Apr 15, 2019, 21:46

*Created originally on Bitbucket by [mattip](https://bitbucket.org/[7.3.7](https://github.com/gitlab-importer/pypy/milestone/7)B267069e3-3102-46be-9577-3d3727467e9c%7D/)*

CPython implemented a special io class to handle utf8 unicode in an interactive terminal for Python3.6. It was accompanied by PEP 528 https://www.python.org/dev/peps/pep-0528/. PyPy does not have this. It is causing the [extra_tests step](http://buildbot.pypy.org/builders/pypy-c-jit-win-x86-32/builds/4550/steps/shell_12/logs/stdio) on the buildbots to fail

**Comment by JamesTheAwesomeDude** [comment]
I have some code that needs to apply a workaround when PEP 528 isn't enacted. \(The workaround is `subprocess.check_output(['chcp.com'], encoding='ascii')`, then using that to look up an *appropriate* value for `sys.stdout.reconfigure(encoding=...)`.\)

Currently, I'm applying this code whenever `platform.system() == 'Windows' and platform.python_implementation() == 'PyPy'`, but this very stupid test will obviously break once PyPy actually does implement PEP 528.

How can I test for PEP 528 support directly, so that my code will not break when PyPy solves this?

**Comment by mattip** [comment]
I'm pretty sure you can get the class name via `sys.stdout.buffer.raw`, which currently shows `<_io._WindowsConsoleIO mode='wb' closefd=False>` on CPython and `<_io.FileIO name='<stdout>' mode='wb' closefd=False>` on PyPy.

**Comment by JamesTheAwesomeDude** [comment]
Interesting. `isinstance(sys.stdout.buffer.raw, io.FileIO)` seems to work as a check, and (at least on CPython) is a pretty robust one, at that.

```bat
python -c "import io, sys; print(type(sys.stdout.buffer.raw), isinstance(sys.stdout.buffer.raw, io.FileIO))" | find "class"
python -c "import io, sys; print(type(sys.stdout.buffer.raw), isinstance(sys.stdout.buffer.raw, io.FileIO))"
```

Are you suggesting that a check like this&mdash;which currently does work, by the way&mdash;**won't** get broken by a PyPy update when PyPy implements similar functionality?:

```python
import platform
import re
import subprocess
import sys

##import colorama

def fix_console_encoding():
  # implementation note: MUST be run before the first read from stdin.
  # (stdout and sterr may be already written-to, albeit maybe corruptedly.)
  if platform.system() == 'Windows':
    ##colorama.just_fix_windows_console()
    if platform.python_implementation() == 'PyPy':
      if isinstance(sys.stdout.buffer.raw, io.FileIO):  # <== Line In Question
        # Workaround for https://github.com/pypy/pypy/issues/2999
        chcp_output = subprocess.check_output(['chcp.com'], encoding='ascii')
        cur_codepage = int(re.match(r'Active code page: (\d+)', chcp_output).group(1))
        cur_encoding = WINDOWS_CODEPAGES[cur_codepage]
        for f in [sys.stdin, sys.stdout, sys.stderr]:
          if f.encoding != cur_encoding
            f.reconfigure(encoding=cur_encoding)
```

**Comment by mattip** [comment]
You don't describe what is the problem you are actually trying to work around. I cannot tell if it is because of `io._WindowsConsoleIO` vs. `io.FileIO` or whether this is some more general encoding issue between PyPy and CPython.

**Comment by JamesTheAwesomeDude** [comment]
> You don't describe what is the problem you are actually trying to work around.

1. PyPy out-of-the-box doesn't use the right encoding for Windows CMD, which causes `print()` to display garbage for non-ASCII characters.
2. I can fix this by using `stdout.reconfigure` after calling a Windows-specific API to get the current console codepage.
3. CPython includes out-of-the-box a different, more aggressive, more low-level, more integrated, more complex fix that totally replaces `stdout` with a synthetic file that's backed by Unicode I/O with Windows-specific APIs.
4. If I apply `reconfigure` on this synthetic file to match the active console codepage, I break *it*
5. I don't know what version of PyPy will fix Unicode printing on Windows.
6. I don't know whether PyPy will use the same solution as CPython to fix Unicode printing on Windows, or a different one.

Because of the conjunction of all of these, I don't know how to write code that will print Unicode characters properly on this version of PyPy *and* all future versions.

If I were guaranteed that you will eventually steal CPython's solution for fixing Unicode output on Windows, then I could just use `isinstance(sys.stdout.buffer.raw, io.FileIO)` as a test.

If I knew exactly what version you'd fix Unicode output on Windows, then I could just use `sys.pypy_version_info < that_version` as a test.

But right now, I don't know what checks \(as far as Python implementation goes\) I should run before calling `stdout.reconfigure(encoding=actual_codepage)` on Windows.

**Comment by JamesTheAwesomeDude** [comment]
Given the `_WindowsConsoleIO` integration difficulties that were posted earlier in this thread, I wouldn't be totally surprised if y'all went with a simpler fix (like setting stdout's encoding to match the active console encoding).

But if I don't know what fix is planned for Unicode printing on Windows, *and* I don't know what version the fix will be included, then I can't ensure that my own code won't break when PyPy eventually fixes Unicode printing on Windows.

**Comment by JamesTheAwesomeDude** [comment]
> Fix for what problem?

> PyPy out-of-the-box doesn't use the right encoding for Windows CMD, which causes print() to display garbage for non-ASCII characters.
>
> I can't ensure that my own code won't break when PyPy eventually fixes Unicode printing on Windows.

**Comment by JamesTheAwesomeDude** [comment]
I'm sorry, but I completely don't understand what you're trying to say, or what you're asking me. I've tried explaining my issue, what I've tried, what my concerns with the existing solutions are, in great detail, but there's some kind of breakdown in communication occurring.

Could you put your question more verbosely \(and explain why the *existing* answers I've given aren't satisfactory\), so maybe I can try to answer it?

**Comment by mattip** [comment]
> PyPy out-of-the-box doesn't use the right encoding for Windows CMD, which causes print() to display garbage for non-ASCII characters

Got it, thanks. That was what I was looking for. What enconding should PyPy be using? Maybe we can solve that first, add a test, and make sure the test passes if we ever re-enable the _WindowsConsoleIO.

**Comment by mattip** [comment]
Ahh, probably we should do this
> set stdout's encoding to match the active console encoding

**Comment by JamesTheAwesomeDude** [comment]
Indeed, there are many possible solutions; configuring stdout to match the console is one option; PEP 528 is a another option, which is incompatible with the first.

But I've been asking if you have any recommendations of how I can tell when PyPy has implemented *any* solution, so that my own code can back off when it's appropriate.

**Comment by JamesTheAwesomeDude** [comment]
It's worth noting that the "cp437" provided by the PSF is *not* identical to the codepage `chcp 437` \(the default codepage on Windows CMD\) implements.

Specifically, the PSF "cp437" encoding fails on exactly these characters:

<pre><code>{u'&#x263a;', u'&#x263b;', u'&#x2665;', u'&#x2666;', u'&#x2663;', u'&#x2660;', u'&#x2642;', u'&#x2640;', u'&#x266b;', u'&#x263c;', u'&#x25ba;', u'&#x25c4;', u'&#x2195;', u'&#x203c;', u'&#xb6;', u'&#xa7;', u'&#x25ac;', u'&#x21a8;', u'&#x2191;', u'&#x2193;', u'&#x2192;', u'&#x2190;', u'&#x221f;', u'&#x25bc;', u'&#x25b2;', u'&#x2302;'}</code></pre>

In the spirit of the historical way Python's handled [WIP, unofficial vendor-spceific codecs](https://github.com/python/cpython/blob/v3.6.0/Lib/encodings/aliases.py#L545), I penciled in "x-dos-cp437" over at [this branch](https://github.com/JamesTheAwesomeDude/pypy/tree/feature/windows-cmd-encoding) in case that will be useful in a "set stdio encoding to actual console encoding" approach.

**Comment by mattip** [comment]
I updated the [winconsole branch](https://github.com/pypy/pypy/commits/winconsole/) so it can translate, and tested it by trying to build and test NumPy. I am still getting an OSError 6 (which indicates an invalid handle) somewhere:
```
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
OSError: 6

```
